smac/intensification/intensification.py

```diff
@@ -137,30 +192,303 @@ def __init__(self, tae_runner: ExecuteTARun,
+            # In this case, we need a branch new challenger
+            if self.stage == IntensifierStage.RUN_FIRST_CONFIG:
+                if incumbent is None:
+                    self.logger.info(
+                        "First run, no incumbent provided;"
+                        " challenger is assumed to be the incumbent"
+                    )
+                    incumbent = challenger
+                else:
+                    inc_runs = run_history.get_runs_for_config(
+                        incumbent,
+                        only_max_observed_budget=True
+                    )
+                    if len(inc_runs) > 0:
+                        self.logger.debug(
+                            "Skipping RUN_FIRST_CONFIG stage since "
+                            "incumbent is already run before"
+                        )
+                        self.stage = IntensifierStage.RUN_INCUMBENT
+
+            # Before we proceed with the intensification loop,
+            # we have to check if it makes sense to run the challenger
+            # That is, if running lines 8-18 but there is no more time due
+            # to adaptive caping or there are no more instances to run
+            # then we can proactively assume we will do a new intensification
+            # round
+            if self.stage in [IntensifierStage.RUN_CHALLENGER, IntensifierStage.RUN_BASIS]:
+                if not self.to_run:
+                    self.to_run, self.inc_sum_cost = self._get_instances_to_run(
+                        incumbent=incumbent,
+                        challenger=challenger,
+                        run_history=run_history,
+                        N=self.N
+                    )
+
+                is_there_time_due_to_adaptive_cap = self._is_there_time_due_to_adaptive_cap(
+                    challenger=challenger,
+                    run_history=run_history,
+                )
+
+                if len(self.to_run) == 0 or not is_there_time_due_to_adaptive_cap:
+
+                    # If no more time, stage transition is a must
+                    if not is_there_time_due_to_adaptive_cap:
+                        self.stage = IntensifierStage.RUN_INCUMBENT
+                        self.logger.debug(
+                            "Stop challenger itensification due "
+                            "to adaptive capping."
+                        )
+
+                    # Nevertheless, if there are no more instances to run,
+                    # we might need to comply with line 17 and keep running the
+                    # same challenger
+                    self.logger.debug("No further runs for challenger possible")
+                    self._process_racer_results(
+                        challenger=challenger,
+                        incumbent=incumbent,
+                        run_history=run_history,
+                    )
+
+                    # Remove the challenger
```

Comment:
Why is the challenger removal actually necessary here? Isn't this done in smbo.py?

Reply:
As per discussed previously, will move to use SKP in favor of more simple code. This will be removed